src/mender-update/daemon/states.cpp

```diff
@@ -800,11 +800,17 @@ void UpdateSaveProvidesState::OnEnter(Context &ctx, sm::EventPoster<StateEvent>
 		artifact_name = artifact.artifact_name;
 	}
 
+	bool deploy_failed = ctx.deployment.failed;
+
+	// Only the artifact_name and group should be committed in the case of a
+	// failing update in order to make this consistent with the old client
+	// behaviour.
 	auto err = ctx.mender_context.CommitArtifactData(
 		artifact_name,
 		artifact.artifact_group,
-		artifact.type_info_provides,
-		artifact.clears_artifact_provides,
+		deploy_failed ? nullopt : optional<context::ProvidesData>(artifact.type_info_provides),
+		deploy_failed ? nullopt
+					  : optional<context::ClearsProvidesData>(artifact.clears_artifact_provides),
```

Comment:
Hmm, I'm not 100% sure, but I think this will *remove* all provides. A missing `ClearsProvidesData` means "clear everything", for backwards compatibility reasons. I think passing in an empty list is better, which means "leave everything alone".